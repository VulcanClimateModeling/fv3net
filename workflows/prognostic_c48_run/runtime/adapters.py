from typing import Mapping, Set, Sequence, Hashable

import xarray as xr

from fv3fit._shared import Predictor

NameDict = Mapping[Hashable, Hashable]


def _invert_dict(d: Mapping) -> Mapping:
    return dict(zip(d.values(), d.keys()))


class RenamingAdapter:
    """Adapter object for renaming model variables

    Attributes:
        model: a model to rename
        rename_in: mapping from standard names to input names of model
        rename_out: mapping from standard names to the output names of model
    
    """

    def __init__(
        self, model: Predictor, rename_in: NameDict, rename_out: NameDict = None
    ):
        self.model = model
        self.rename_in = rename_in
        self.rename_out = {} if rename_out is None else rename_out

    def _rename(self, ds: xr.Dataset, rename: NameDict) -> xr.Dataset:

        all_names = set(ds.dims) & set(rename)
        rename_restricted = {key: rename[key] for key in all_names}
        redimed = ds.rename_dims(rename_restricted)

        all_names = set(ds.data_vars) & set(rename)
        rename_restricted = {key: rename[key] for key in all_names}
        return redimed.rename(rename_restricted)

    def _rename_inputs(self, ds: xr.Dataset) -> xr.Dataset:
        return self._rename(ds, self.rename_in)

    def _rename_outputs(self, ds: xr.Dataset) -> xr.Dataset:
        return self._rename(ds, _invert_dict(self.rename_out))

    @property
    def input_variables(self) -> Set[str]:
        invert_rename_in = _invert_dict(self.rename_in)
        return {invert_rename_in.get(var, var) for var in self.model.input_variables}

    def predict(self, arg: xr.Dataset) -> xr.Dataset:
        input_ = self._rename_inputs(arg)
        prediction = self.model.predict(input_)
        return self._rename_outputs(prediction)


class StackingAdapter:
    """Wrap a model to work with unstacked inputs
    
    Attributes
        model: a model to stack
        sample_dims: sample dimension name of the resulting stacked array
    
    """

    def __init__(self, model: Predictor, sample_dims: Sequence[str]):
        self.model = model
        self.sample_dims = sample_dims

    @property
    def input_variables(self) -> Set[str]:
        return set(self.model.input_variables)

    def predict(self, ds: xr.Dataset) -> xr.Dataset:
        self.model.predict_columnwise(ds, sample_dims=self.sample_dims)
