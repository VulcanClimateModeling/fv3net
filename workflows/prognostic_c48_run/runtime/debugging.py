import pdb
from mpi4py import MPI
import xarray as xr


def breakpoint(rank=0):
    if rank == MPI.COMM_WORLD.rank:
        pdb.set_trace()


def checkpoint(state, path, rank=0):
    if MPI.COMM_WORLD.rank == 0:
        variables = state.checkpoint()
        input_ = xr.Dataset({key: variables[key] for key in variables})
        input_.to_netcdf(path)
