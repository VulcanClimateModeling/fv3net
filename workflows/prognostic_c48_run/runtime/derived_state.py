import cftime
from typing import Mapping, Hashable
import xarray as xr

import fv3gfs.util
from vcm import DerivedMapping


class FV3StateMapper:
    """ A mapping interface for the FV3GFS getter.
        
    Maps variables to the common names used in shared functions.
    By default adds mapping {"lon": "longitude", "lat": "latitude"}
    """

    def __init__(self, getter, alternate_keys: Mapping[str, str] = None):
        self._getter = getter
        self._alternate_keys = alternate_keys or {"lon": "longitude", "lat": "latitude"}

    def __getitem__(self, key: str) -> xr.DataArray:
        if key == "time":
            time = self._getter.get_state(["time"])["time"]
            return xr.DataArray(time, dims=["time"], coords={"time": [time]})
        elif key == "latent_heat_flux":
            return self._getter.get_diagnostic_by_name("lhtfl").data_array
        elif key == "total_water":
            return self._total_water()
        else:
            if key in self._alternate_keys:
                key = self._alternate_keys[key]
            return self._getter.get_state([key])[key].data_array

    def _total_water(self):
        a = self._getter.get_tracer_metadata()
        water_species = [name for name in a if a[name]["is_water"]]
        return sum(self[name] for name in water_species)


class DerivedFV3State:
    """A uniform mapping-like interface to the FV3GFS model state
    
    This class wraps the fv3gfs getters with the FV3StateMapper, that always returns
    DataArray and has time as an attribute (since this isn't a DataArray).
    
    This insulates runfiles from the details of Quantity
    
    """

    def __init__(self, getter):
        """
        Args:
            getter: the fv3gfs object or a mock of it.
        """
        self._getter = getter
        self._mapper = DerivedMapping(FV3StateMapper(getter, alternate_keys=None))

    @property
    def time(self) -> cftime.DatetimeJulian:
        return self._getter.get_state(["time"])["time"]

    def __getitem__(self, key: Hashable) -> xr.DataArray:
        return self._mapper[key]

    def __setitem__(self, key: str, value: xr.DataArray):
        self._getter.set_state_mass_conserving(
            {key: fv3gfs.util.Quantity.from_data_array(value)}
        )

    def update(self, items: Mapping[Hashable, xr.DataArray]):
        """Update state from another mapping

        This may be faster than setting each item individually.
        
        Same as dict.update.
        """
        self._getter.set_state_mass_conserving(
            {
                key: fv3gfs.util.Quantity.from_data_array(value)
                for key, value in items.items()
            }
        )
