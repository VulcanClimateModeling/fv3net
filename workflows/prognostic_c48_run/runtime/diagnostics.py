from typing import (
    Any,
    Sequence,
    Container,
    Mapping,
    List,
    Union,
    MutableMapping,
    Hashable,
)
import datetime
import cftime
import logging
from vcm.catalog import catalog
from vcm.cubedsphere import center_and_rotate_xy_winds
import fv3gfs.util

import xarray as xr

logger = logging.getLogger(__name__)

TEMP = "air_temperature"
SPHUM = "specific_humidity"
DELP = "pressure_thickness_of_atmospheric_layer"
PRECIP_RATE = "surface_precipitation_rate"
cp = 1004
gravity = 9.81

State = MutableMapping[Hashable, xr.DataArray]


class All(Container):
    """A container that contains every thing
    
    This is useful for cases where we want an ``in`` check to always return True.

    Example:
        >>> all = All()
        >>> 'x' in all
        True
        >>> 1232.1 in all
        True
    """

    def __contains__(self, value: Any) -> bool:
        return True


class SelectedTimes(Container[cftime.DatetimeJulian]):
    TIME_FMT: str = r"%Y%m%d.%H%M%S"

    def __init__(self, times=Sequence[str]):
        self._time_stamps = times

        # see if there is an error
        self.times

    @property
    def _times(self) -> Sequence[datetime.datetime]:
        return [
            datetime.datetime.strptime(time, self.TIME_FMT)
            for time in self._time_stamps
        ]

    @property
    def times(self) -> Sequence[cftime.DatetimeJulian]:
        return [cftime.DatetimeJulian(*time.timetuple()) for time in self._times]

    def __contains__(self, time: cftime.DatetimeJulian) -> bool:
        return time in self.times


class IntervalTimes(Container[cftime.DatetimeJulian]):
    def __init__(
        self, frequency_seconds: Union[float, int], initial_time: cftime.DatetimeJulian,
    ):
        """
        Args:
            frequency_seconds: the output frequency from the initial time
            initial_time: the initial time to start the period
            
        """
        self._frequency_seconds = frequency_seconds
        self.initial_time = initial_time
        if self.frequency > datetime.timedelta(days=1.0) and initial_time is None:
            raise ValueError(
                "Minimum output frequency is daily when intial_time is not provided."
            )

    @property
    def frequency(self) -> datetime.timedelta:
        return datetime.timedelta(seconds=self._frequency_seconds)

    def __contains__(self, time) -> bool:
        time_since_initial_time = time - self.initial_time
        quotient = time_since_initial_time % self.frequency
        return quotient == datetime.timedelta(seconds=0)


def _assign_units_if_none_present(array: xr.DataArray, units=None):
    return array.assign_attrs(units=array.attrs.get("units", units))


class DiagnosticFile:
    """A object representing a diagnostics file

    Provides a similar interface as the "diag_table"

    Replicates the abilities of the fortran models's diag_table by allowing
    the user to specify different output times for distinct sets of
    variables.
    """

    def __init__(
        self,
        monitor: fv3gfs.util.ZarrMonitor,
        times: Container[cftime.DatetimeJulian],
        variables: Container,
    ):
        """
        Args:
            monitor: an underlying monitor to store the data in
            times: the set of times (potentially infinite) to save the data at
            variables: a container of variables to save

        Note:

            The containers used for times and variables do not need to be
            concrete lists or python sequences. They only need to satisfy the
            abstract ``Container`` interface. Please see the special
            containers for outputing times above:

            - ``IntervalTimes``
            - ``SelectedTimes``

            as well as the generic ``All`` container that contains the entire
            Universe!
        """
        self._monitor = monitor
        self.times = times
        self.variables = variables

    def observe(
        self, time: cftime.DatetimeJulian, diagnostics: Mapping[str, xr.DataArray]
    ):
        """Possibly store the data into the monitor
        """
        if time in self.times:
            quantities = {
                # need units for from_data_array to work
                key: fv3gfs.util.Quantity.from_data_array(
                    _assign_units_if_none_present(diagnostics[key], "unknown")
                )
                for key in diagnostics
                if key in self.variables
            }

            # patch this in manually. the ZarrMonitor needs it.
            # We should probably modify this behavior.
            quantities["time"] = time
            self._monitor.store(quantities)


def _get_times(
    d, initial_time: cftime.DatetimeJulian
) -> Container[cftime.DatetimeJulian]:
    kind = d.get("kind", "every")
    if kind == "interval":
        return IntervalTimes(d["frequency"], initial_time)
    elif kind == "selected":
        return SelectedTimes(d["times"])
    elif kind == "every":
        return All()
    else:
        raise NotImplementedError(f"Time {kind} not implemented.")


def _config_to_diagnostic_file(
    diag_file_config: Mapping, partitioner, comm, initial_time: cftime.DatetimeJulian,
) -> DiagnosticFile:
    monitor = fv3gfs.util.ZarrMonitor(
        diag_file_config["name"], partitioner, mpi_comm=comm
    )
    return DiagnosticFile(
        monitor=monitor,
        variables=diag_file_config.get("output_variables", All()),
        times=_get_times(diag_file_config.get("times", {}), initial_time),
    )


def get_diagnostic_files(
    config: Mapping,
    partitioner: fv3gfs.util.CubedSpherePartitioner,
    comm,
    initial_time: cftime.DatetimeJulian,
) -> List[DiagnosticFile]:
    """Initialize a list of diagnostic file objects from a configuration dictionary
    Note- the default here is to save all the variables in the diagnostics.
    The default set of variables can be overwritten by inserting a default diagnostics
    config entry for each runfile, e.g. ../prepare_config.py does this for
    the sklearn runfile.

    Args:
        config: A loaded "fv3config" dictionary with a "diagnostics" section
        paritioner: a partioner object used for writing, maybe it would be
            cleaner to pass a factory
        comm: an MPI Comm object
        initial_time: the initial time of the simulation.

    """
    diag_configs = config.get("diagnostics", [])
    if len(diag_configs) > 0:
        return [
            _config_to_diagnostic_file(config, partitioner, comm, initial_time)
            for config in diag_configs
        ]
    else:
        # Keep old behavior for backwards compatiblity
        output_name = config["scikit_learn"]["zarr_output"]
        default_config = {"name": output_name, "times": {}, "variables": All()}
        return [
            _config_to_diagnostic_file(default_config, partitioner, comm, initial_time)
        ]


def compute_ml_diagnostics(state, ml_tendency):

    net_moistening = (ml_tendency["dQ2"] * state[DELP] / gravity).sum("z")
    physics_precip = state[PRECIP_RATE]

    return dict(
        air_temperature=state[TEMP],
        specific_humidity=state[SPHUM],
        pressure_thickness_of_atmospheric_layer=state[DELP],
        net_moistening=(net_moistening)
        .assign_attrs(units="kg/m^2/s")
        .assign_attrs(description="column integrated ML model moisture tendency"),
        net_heating=(ml_tendency["dQ1"] * state[DELP] / gravity * cp)
        .sum("z")
        .assign_attrs(units="W/m^2")
        .assign_attrs(description="column integrated ML model heating"),
        water_vapor_path=(state[SPHUM] * state[DELP] / gravity)
        .sum("z")
        .assign_attrs(units="mm")
        .assign_attrs(description="column integrated water vapor"),
        physics_precip=(physics_precip)
        .assign_attrs(units="kg/m^2/s")
        .assign_attrs(
            description="surface precipitation rate due to parameterized physics"
        ),
    )


def compute_ml_momentum_diagnostics(state, tendency):
    delp = state[DELP]

    dQu = tendency.get("dQu", xr.zeros_like(delp))
    dQv = tendency.get("dQv", xr.zeros_like(delp))
    column_integrated_dQu = _mass_average(dQu, delp, "z")
    column_integrated_dQv = _mass_average(dQv, delp, "z")
    return dict(
        column_integrated_dQu=column_integrated_dQu.assign_attrs(
            units="m s^-2",
            description="column integrated zonal wind tendency due to ML",
        ),
        column_integrated_dQv=column_integrated_dQv.assign_attrs(
            units="m s^-2",
            description="column integrated meridional wind tendency due to ML",
        ),
    )


def rename_diagnostics(diags):
    """Postfix ML output names with _diagnostic and create zero-valued outputs in
    their stead. Function operates in place."""
    ml_tendencies = {
        "net_moistening",
        "net_heating",
        "column_integrated_dQu",
        "column_integrated_dQv",
    }
    ml_tendencies_in_diags = ml_tendencies & set(diags)
    for variable in ml_tendencies_in_diags:
        attrs = diags[variable].attrs
        diags[f"{variable}_diagnostic"] = diags[variable].assign_attrs(
            description=attrs["description"] + " (diagnostic only)"
        )
        diags[variable] = xr.zeros_like(diags[variable]).assign_attrs(attrs)


def compute_nudging_diagnostics(
    state: State, nudging_tendency: State, label: str = "_tendency_due_to_nudging"
):
    """
    Compute diagnostic variables for nudging"""

    net_moistening = (
        (nudging_tendency[SPHUM] * state[DELP] / gravity)
        .sum("z")
        .assign_attrs(units="kg/m^2/s")
        .assign_attrs(description="column integrated moistening due to nudging")
    )
    net_heating = (
        (nudging_tendency[TEMP] * state[DELP] / gravity * cp)
        .sum("z")
        .assign_attrs(units="W/m^2")
        .assign_attrs(description="column integrated heating due to nudging")
    )
    water_vapor_path = (
        (state[SPHUM] * state[DELP] / gravity)
        .sum("z")
        .assign_attrs(units="mm")
        .assign_attrs(description="column integrated water vapor")
    )
    physics_precip = (
        state[PRECIP_RATE]
        .assign_attrs(units="kg/m^2/s")
        .assign_attrs(
            description="surface precipitation rate due to parameterized physics"
        )
    )

    diags = dict(
        net_moistening_due_to_nudging=net_moistening,
        net_heating_due_to_nudging=net_heating,
        water_vapor_path=water_vapor_path,
        physics_precip=physics_precip,
    )

    if ("x_wind" in nudging_tendency.keys()) and ("y_wind" in nudging_tendency.keys()):
        wind_rotation_matrix = catalog["wind_rotation/c48"].to_dask()
        u_tendency, v_tendency = center_and_rotate_xy_winds(
            wind_rotation_matrix, nudging_tendency["x_wind"], nudging_tendency["y_wind"]
        )
        rotation_mapping = {
            ("u-wind", "x_wind"): u_tendency,
            ("v-wind", "y_wind"): v_tendency,
        }
        for names, tendency in rotation_mapping.items():
            a_name, d_name = names
            integrated_wind_tendency = _mass_average(tendency, state[DELP], "z")
            diags[
                f"column_integrated_{a_name}_tendency_due_to_nudging"
            ] = integrated_wind_tendency.assign_attrs(
                units="m s^-2",
                description=(
                    f"column mass-averaged {a_name} wind tendency due to nudging"
                ),
            )

    if DELP in nudging_tendency:
        net_mass_tendency = (
            (nudging_tendency[DELP] / gravity)
            .sum("z")
            .assign_attrs(
                units="kg/m^2/s",
                description="column_integrated mass tendency due to nudging",
            )
        )
        diags["net_mass_tendency_due_to_nudging"] = net_mass_tendency
    diags.update(_append_key_label(nudging_tendency, label))

    return diags


def _append_key_label(d, suffix):
    return_dict = {}
    for key, value in d.items():
        return_dict[key + suffix] = value
    return return_dict


def _mass_average(da: xr.DataArray, delp: xr.DataArray, vertical_dim: str = "z"):
    total_thickness = delp.sum(vertical_dim)
    mass_average = (da * delp).sum(vertical_dim) / total_thickness
    mass_average.assign_attrs(**da.attrs)
    return mass_average
