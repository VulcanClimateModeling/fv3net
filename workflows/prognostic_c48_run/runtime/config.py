from typing import List, Optional
import dataclasses
import yaml
import f90nml

import dacite

from runtime.diagnostics.manager import (
    DiagnosticFileConfig,
    FortranFileConfig,
    get_chunks,
)
from runtime.steppers.nudging import NudgingConfig
from runtime.steppers.machine_learning import MachineLearningConfig

FV3CONFIG_FILENAME = "fv3config.yml"


@dataclasses.dataclass
class UserConfig:
    """The top-level object for python runtime configurations

    Attributes:
        diagnostics: list of diagnostic file configurations
        fortran_diagnostics: list of Fortran diagnostic outputs. Currently only used by
            post-processing and so only name and chunks items need to be specified.
        scikit_learn: a machine learning configuration
        nudging: nudge2fine configuration. Cannot be used if any scikit_learn model
            urls are specified.
        step_tendency_variables: variables to compute the tendencies of.
            These could in principle be inferred from the requested diagnostic
            names.
        step_storage_variables: variables to compute the storage of. Needed for certain
            diagnostics.
    """

    diagnostics: List[DiagnosticFileConfig]
    fortran_diagnostics: List[FortranFileConfig]
    scikit_learn: MachineLearningConfig = MachineLearningConfig()
    nudging: Optional[NudgingConfig] = None
    step_tendency_variables: List[str] = dataclasses.field(
        default_factory=lambda: list(
            ("specific_humidity", "air_temperature", "eastward_wind", "northward_wind",)
        )
    )
    step_storage_variables: List[str] = dataclasses.field(
        default_factory=lambda: list(("specific_humidity", "total_water"))
    )


def get_config() -> UserConfig:
    """Open the configurations for this run
    
    .. warning::
        Only valid at runtime
    """
    with open("fv3config.yml") as f:
        config = yaml.safe_load(f)
    return dacite.from_dict(UserConfig, config)


def get_namelist() -> f90nml.Namelist:
    """Open the fv3 namelist
    
    .. warning::
        Only valid at runtime
    """
    return f90nml.read("input.nml")


def write_chunks(config: UserConfig):
    """Given UserConfig, write chunks to 'chunks.yaml'"""
    diagnostic_file_configs = (
        config.fortran_diagnostics + config.diagnostics  # type: ignore
    )
    chunks = get_chunks(diagnostic_file_configs)
    with open("chunks.yaml", "w") as f:
        yaml.safe_dump(chunks, f)
