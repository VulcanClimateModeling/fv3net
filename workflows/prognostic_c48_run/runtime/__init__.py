from .config import get_namelist, get_config
from .capture import capture_stream, capture_stream_mpi, capture_fv3gfs_funcs
from .diagnostics import (
    get_diagnostic_files,
    DiagnosticFile,
    compute_ml_diagnostics,
    compute_ml_momentum_diagnostics,
    compute_nudging_diagnostics,
    rename_diagnostics,
    default_diagnostics,
)
from .adapters import RenamingAdapter, MultiModelAdapter
from .derived_state import DerivedFV3State
from .nudging import (
    nudging_timescales_from_dict,
    setup_get_reference_state,
    get_nudging_tendency,
    set_state_sst_to_reference,
)

__version__ = "0.1.0"
