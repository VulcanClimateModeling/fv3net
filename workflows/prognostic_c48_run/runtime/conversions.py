import xarray as xr
import pace.util

from runtime.types import QuantityState


def _insert_units_if_necessary(da: xr.DataArray) -> xr.DataArray:
    if "units" not in da.attrs:
        return da.assign_attrs(units="")
    else:
        return da


def dataset_to_quantity_state(ds: xr.Dataset) -> QuantityState:
    quantity_state: QuantityState = {
        variable: pace.util.Quantity.from_data_array(
            _insert_units_if_necessary(ds[variable])
        )
        for variable in ds.data_vars
    }
    return quantity_state


def quantity_state_to_dataset(quantity_state: QuantityState) -> xr.Dataset:
    ds = xr.Dataset(
        {
            variable: quantity_state[variable].data_array
            for variable in quantity_state.keys()
        }
    )
    return ds
