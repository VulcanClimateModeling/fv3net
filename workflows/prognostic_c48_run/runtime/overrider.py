import dataclasses
import functools
import logging
from typing import Hashable, Mapping, MutableMapping, Set

import cftime
import fsspec
import xarray as xr

import fv3gfs.util
from runtime.monitor import Monitor
from runtime.types import Diagnostics, Step
from runtime.derived_state import DerivedFV3State
from runtime.utils import quantity_state_to_ds, ds_to_quantity_state

QuantityState = MutableMapping[Hashable, fv3gfs.util.Quantity]

logger = logging.getLogger(__name__)


@dataclasses.dataclass
class OverriderConfig:
    """Configuration for overriding tendencies from a step.
    
    Attributes:
        url: path to dataset containing tendencies.
        variables: mapping from state name to name of corresponding tendency in
            provided dataset. For example: {"air_temperature": "fine_res_Q1"}.
    """

    url: str
    variables: Mapping[str, str]


@dataclasses.dataclass
class OverriderAdapter:
    """Wrap a Step function to override certain tendencies."""

    config: OverriderConfig
    state: DerivedFV3State
    communicator: fv3gfs.util.CubedSphereCommunicator
    timestep: float
    diagnostic_variables: Set[str] = dataclasses.field(default_factory=set)

    def __post_init__(self: "OverriderAdapter"):
        if self.communicator.rank == 0:
            logger.debug(f"Opening tendency overriding dataset from: {self.config.url}")
        tile = self.communicator.partitioner.tile_index(self.communicator.rank)
        mapper = fsspec.get_mapper(self.config.url)
        ds = xr.open_zarr(mapper, consolidated=True)
        ds = ds[list(self.config.variables.values())].isel(tile=tile)
        self._tendency_ds = DatasetCachedByChunk(ds, "time")

    def _open_tendencies_dataset(self, time: cftime.DatetimeJulian) -> xr.Dataset:
        ds = xr.Dataset()
        if self.communicator.tile.rank == 0:
            ds = self._tendency_ds.load(time).drop_vars("time")
        tendencies = self.communicator.tile.scatter_state(ds_to_quantity_state(ds))
        return quantity_state_to_ds(tendencies)

    @property
    def monitor(self) -> Monitor:
        return Monitor.from_variables(
            self.diagnostic_variables, self.state, self.timestep
        )

    def _override(self, name: str, func: Step) -> Diagnostics:
        if self.communicator.rank == 0:
            logger.debug(f"Overriding tendencies for {name}.")
        tendencies = self._open_tendencies_dataset(self.state.time)
        before = self.monitor.checkpoint()
        diags = func()
        change_due_to_func = self.monitor.compute_change(name, before, self.state)
        for variable_name, tendency_name in self.config.variables.items():
            with xr.set_options(keep_attrs=True):
                self.state[variable_name] = (
                    before[variable_name] + tendencies[tendency_name] * self.timestep
                )
        change_due_to_overriding = self.monitor.compute_change(
            "override", before, self.state
        )
        return {**diags, **change_due_to_func, **change_due_to_overriding}

    def __call__(self, name: str, func: Step) -> Step:
        """Override tendencies from a function that updates the State.
        
        Args:
            name: a label for the step that is being overriden.
            func: a function that updates the State and return Diagnostics.
            
        Returns:
            overridden_func: a function which observes the change to State
            done by ``func`` and overrides the change for specified variables.
        """

        def step() -> Diagnostics:
            return self.override(name, func)

        step.__name__ = func.__name__
        return step


class DatasetCachedByChunk:
    def __init__(self, ds: xr.Dataset, dim: str):
        """Wrapper to cache the most recently loaded chunks of an xarray dataset.
        
        Args:
            ds: A chunked xarray dataset.
            dim: Aimension along which different chunks are being accessed.

        Example:
            >>> cached_ds = DatasetCachedByChunk(ds, "time")
            >>> cached_ds.load("2016-08-01T00:00:00")  # slow
            >>> cached_ds.load("2016-08-01T00:15:00")  # fast if on same chunk
        """
        self._ds = ds
        self._dim = dim
        self._chunksize = ds.chunks[dim][0]
        self._index = ds.get_index(dim)

    def load(self, key):
        """Load key from the given dimension of dataset."""
        idx = self._index.get_loc(key)
        chunknum, chunk_idx = divmod(idx, self._chunksize)
        return self._load_chunk(chunknum).isel({self._dim: chunk_idx})

    @functools.lru_cache(2)
    def _load_chunk(self, i):
        chunk_slice = slice(i * self._chunksize, (i + 1) * self._chunksize)
        return self._ds.isel({self._dim: chunk_slice}).load()
