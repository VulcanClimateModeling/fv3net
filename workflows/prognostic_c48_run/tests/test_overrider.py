import numpy as np
import xarray as xr

from runtime.overrider import DatasetCachedByChunk

da = xr.DataArray(
    np.arange(11 * 4 * 6).reshape((11, 4, 6)),
    dims=["time", "x", "y"],
    coords={"time": np.arange(2000, 2011)},
).chunk({"time": 5, "x": 4, "y": 4})
ds = xr.Dataset({"var": da})


def test_dataset_caching():
    cached = DatasetCachedByChunk(ds, "time")
    ds_year_2000 = cached.load(2000)
    xr.testing.assert_identical(ds_year_2000, ds.sel(time=2000))
    assert cached._load_chunk.cache_info().hits == 0
    cached.load(2001)
    assert cached._load_chunk.cache_info().hits == 1

