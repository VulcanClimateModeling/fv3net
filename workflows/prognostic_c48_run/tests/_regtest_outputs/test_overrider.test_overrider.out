some_diag 0b0a9008ecec306ce21cef94196b8399
storage_of_mass_due_to_add_one 00c9935733a45000e0c5c8d9b1833f21
storage_of_mass_due_to_override 00c9935733a45000e0c5c8d9b1833f21
tendency_of_air_temperature_due_to_add_one a98d20e04817a73a6ed0360d5c145cf1
tendency_of_air_temperature_due_to_override bb514b470c45b6b1dd05ee119aea4399
